### PR TITLE
[XF] EventToCommandBehavior

### PR DESCRIPTION
**Pull request by joacar**
Copied the `BehaviorsBase<T>` from pr for multipagebehavior.

PR for #714

### CODE REVIEW

**Comment by brianlagunas** [comment]
You can either wait until we get the base behavior infrastructure in place, and update this PR later, or close this PR and wait to update your PR then resubmit.  Also, I think we would like to avoid the event args converter and use a parameter path if possible.

**Comment by joacar** [comment]
I thought I wait and then just merge/rebase. Good idea. That is more inline with `InvokeCommandAction`

**Comment by brianlagunas** [comment]
I am almost leaning towards naming it InvokeCommandAction just to align with WPF, since it is the same thing, but I prefer the EventToCommand name.  Better yet, I may just rename the WPF version to EventToCommand :)

**Comment by joacar** [comment]
Hm `EventArgsParameterPath` - I don't like that.. It is about the property/member of event args so perhaps `EventArgs{Member,Property}Path`. Or should we just stick with the WPF property name TriggerParameterPath?

I'll remove the  `EventArgsConverter` and  `EventArgsConverterParameter` and update the example and documentation.

**Comment by powerdude** [comment]
i don't think you should remove the converter as i've found it useful in other implementations of this feature in another library

**Comment by dansiegel** [comment]
As long as you're not a waitress giving me options for sides with my meal... I am a HUGE fan of options...  I would rather see this provide you the ability to provide a converter or path. Also I would think that overriding this entirely so that you can just bind a CommandParameter and pass it directly in would be awesome...

**Comment by joacar** [comment]
:) Yes I dont mind it either. Should the `CommandParameter` take precedens over the others?

Order of evulation

1. `CommandParameter` set then execute
2. `EventArgsParameterPath` set then evaluate member and execute
3.  `EventArgsConverter` set then evaluate converter and execute

Also I will evaluate the `CanExecute` earlier to prevent unecessary work if the command isn't even to be executed.

**Comment by joacar** [comment]
@brianlagunas When time comes to merge this I think you should consider [merge-n-squash](https://help.github.com/articles/about-pull-request-merge-squashing/) to prevent the large amount of commits to be present in history for **master** branch. This way only one commit wil be visible and that is the merge commit

**Comment by dansiegel** [comment]
thinking about this a little more here the default for the ```CommandParameter``` should be null. Given that I would maybe expect it to work as follows:

```cs
if( CommandParameter == null )
{
    if( Converter != null )
    {
        CommandParameter = Converter.Convert(...);
    }
    else if( !string.IsNullEmptyOrWhitespace( Path ) )
    {
        // Set CommandParameter
    }
}

if( Command?.CanExecute( CommandParameter ) )
{
    Command.Execute( CommandParameter );
}
```

**Comment by powerdude** [comment]
i don't think you want to set the `CommandParameter` property or you'll set a property the user never set and this code wouldn't fire a second time.  It should be saved to a local variable.

**Comment by dansiegel** [comment]
valid point.

**Comment by joacar** [comment]
The way it works now is, if I'm not missing something, the way you describe it @dansiegel 

````c#
var parameter = CommandParameter;
if(parameter == null && !string.IsNullOrEmpty(Path))
{
    // This will only be executed if CommandParameter is not set and Path is set
     parameter = EvaluatePathProperty();
}
if(parameter == null && EventArgsConverter != null)
{
    // This will only be executed if parameter is not set (either CommandParameter is not set or Path is not set or Path has not evaluated to any non-null property).
    parameter = EventArgsConverter.Convert(...);
}

````

**Comment by dansiegel** [comment]
@joacar that looks good. didn't realize you had updated it since your earlier post.

File: Source/Xamarin/Prism.Forms.Tests/Behaviors/EventToCommandBehaviorFixture.cs [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+        public void Command_OrderOfExecution()
+        {
+            const string commandParameter = "ItemProperty";
+            var converter = new ItemTappedEventArgsConverter(false);
+            var behavior = new EventToCommandBehaviorMock
+            {
+                EventName = "ItemTapped",
+                EventArgsConverter = converter,
+                CommandParameter = commandParameter,
+                Command = new Command(o =>
```
Comment (brianlagunas) [inline Source/Xamarin/Prism.Forms.Tests/Behaviors/EventToCommandBehaviorFixture.cs]: Can you update the Commands to use Prism DelegateCommand instead?

File: Documentation/Xamarin.Forms/6-EventToCommandBehavior.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
+﻿# Using the EventToCommandBehavior
+
+The `EventToCommandBehavior` class provide a convenient way to, in XAML, "bind" events to `ICommand` according to MVVM paradigm to avoid code behind.
+
+## Usage
+
+The `EventToCommandBehavior` expose the following properties
```
Comment (brianlagunas) [inline Documentation/Xamarin.Forms/6-EventToCommandBehavior.md]: Can you add the EventArgsParameterPath and a sample on how it's used?

**Comment by joacar** [comment]
Should we support changing the **EventName** from the view model? E.g. start with _Focused_ and when the command is executed just listen to _Unfocused_ event.

**Comment by brianlagunas** [comment]
So I finally started looking at this, and I have a couple of thoughts going through my head.  First, we have this same type of behavior in WPF, except it's called InvokeCommandAction.  In WPF, you can assign multiple InvokeCommandActions to an event.  Then I started thinking, wait, I am sure this has already been written, and then I found David Britch's Behaviors for Xamarin.Forms that does this exact same thing.  

https://github.com/davidbritch/behaviors

Why should we include it in Prism, if it is already available in another package?

**Comment by dansiegel** [comment]
@brianlagunas I believe the argument here is that the framework provided by Prism is heavily dependent on Commands and Prism should provide this behavior out of the box rather than requiring a dev to either implement it or include yet another library.

**Comment by brianlagunas** [comment]
Well I'm not sure that's a strong argument.  Another argument is that everything shouldn't be in one assembly, instead these should be separated out into another assembly and I can chose to bring them in is I want.

Plus, I was trying to point out the limitations regarding EventToCommand vs. the new implementation of the concept in the link I provided.  Essentially EventToCommand has been replaced with a much more flexible and feature rich solution similar to that of WPF.

**Comment by joacar** [comment]
The implementation in the link is a better approach then this, I agree on that. 

The argument " ... instead these should be separated out into another assembly ..." - does that say the implementation of "EventToCommand" in WPF should be removed from Prism and made a available as a Nuget?

I think that "event to command" is a common issue that a developer meet trying to follow MVVM pattern and hence Prism, as a good MVVM framework, should provide the toolset to handle this.

**Comment by brianlagunas** [comment]
Alright, I have decided that we will keep it simple and only provide a simple EventToCommand behavior.  I don't want Prism to become a do everything framework, so if anyone requires a more custom implementation to execute multiple behaviors for the same event, then they will need to use their own implementation or an existing OSS behavior framework such as David's.

Can you please fix the merge conflicts and get the PR building again?

**Comment by JerryNixon** [comment]
@brianlagunas I could not agree with you more. It's endlessly tempting to build the abstraction infrastructure and do everything with a framework that has a very explicit purpose otherwise. Keep it simple. Keep it clean. And keep your goals as few as possible. We've been backed into a corner with some things in Template 10. You definitely don't want to be there.

**Comment by brianlagunas** [comment]
This has been merged via #888

**Pull request closed without merging**